packages/react-ui/src/app/features/builder/flow-canvas/context-menu/canvas-context-menu-content.tsx

```diff
@@ -60,59 +67,126 @@ export const CanvasContextMenuContent = ({
+
   const showCopy =
     showCopyPaste &&
     !doSelectedNodesIncludeTrigger &&
     contextMenuType === ContextMenuType.STEP;
 
+  const { onPaste } = usePaste();
+
   return (
     <>
       {showCopy && (
-        <ContextMenuItem disabled={disabled} onClick={copySelectedArea}>
+        <ContextMenuItem
+          disabled={disabled}
+          onClick={() => {
+            if (selectedStep) {
+              const step = flowHelper.getStep(flowVersion, selectedStep);
+              copyAction(step as Action);
+              return;
+            }
+
+            copySelectedArea();
+          }}
+        >
           <ShortcutWrapper shortcut={CanvasShortcuts['Copy']}>
             <Copy className="w-4 h-4"></Copy> {t('Copy')}
           </ShortcutWrapper>
         </ContextMenuItem>
       )}
 
       <>
-        {showPasteAfterLastStep && showCopyPaste && (
+        {showPasteAfterLastStep && (
           <ContextMenuItem
             disabled={disabledPaste}
-            onClick={() => {
-              // // https://linear.app/openops/issue/OPS-854/add-paste-logic
-            }}
+            onClick={() =>
+              onPaste(
+                actionToPaste as Action,
+                StepLocationRelativeToParent.AFTER,
+                selectedStep,
+              )
+            }
             className="flex items-center gap-2"
           >
-            <ClipboardPlus className="w-4 h-4"></ClipboardPlus>{' '}
+            <ClipboardPlus className="w-4 h-4"></ClipboardPlus>
             {t('Paste After Last Step')}
           </ContextMenuItem>
         )}
         {showPasteAsFirstLoopAction && (
+          <ContextMenuItem
+            disabled={disabledPaste}
+            onClick={() =>
+              onPaste(
+                actionToPaste as Action,
+                StepLocationRelativeToParent.INSIDE_LOOP,
+                selectedStep,
+              )
+            }
+            className="flex items-center gap-2"
+          >
+            <ClipboardPaste className="w-4 h-4"></ClipboardPaste>
+            {t('Paste inside Loop')}
+          </ContextMenuItem>
+        )}
+        {showPasteInConditionBranch && (
+          <ContextMenuItem
+            disabled={disabledPaste}
+            onClick={() =>
+              onPaste(
+                actionToPaste as Action,
+                StepLocationRelativeToParent.INSIDE_TRUE_BRANCH,
+                selectedStep,
+              )
+            }
+            className="flex items-center gap-2"
+          >
+            <ClipboardPaste className="w-4 h-4"></ClipboardPaste>
+            {t('Paste inside first branch')}
+          </ContextMenuItem>
+        )}
+        {showPasteInSplitBranch && (
           <ContextMenuItem
             disabled={disabledPaste}
             onClick={() => {
-              // https://linear.app/openops/issue/OPS-854/add-paste-logic
+              const branchNodeId = firstSelectedStep?.settings.options[0].id;
```

Comment:
```suggestion
              const branchNodeId = firstSelectedStep.settings.options[0].id;
```
in this case `firstSelectedStep` is always truthy because   
```
const showPasteInSplitBranch =
    contextMenuType === ContextMenuType.STEP &&
    firstSelectedStep?.type === ActionType.SPLIT;
```